### PR TITLE
Remove http client leak

### PR DESCRIPTION
**Pull request by i386**
The ClosableHttpClient is never released correctly if an exception occurs. This could lead to leaked unclosed connections to Github.

Note: this PR depends on #61 and needs a rebase when it is merged. As such it contains all the improvements in #62 and this PR.

PTAL @samrocketman

### CODE REVIEW

File: src/main/java/org/jenkinsci/plugins/GithubSecurityRealm.java [hunk 1]
```diff
@@ -478,6 +403,33 @@ public HttpResponse doFinishLogin(StaplerRequest request)
         return HttpResponses.redirectToContextRoot();   // referer should be always there, but be defensive
     }
 
+    @Nullable
+    private String getAccessToken(@Nonnull String code) throws IOException {
```
Comment (i386) [inline src/main/java/org/jenkinsci/plugins/GithubSecurityRealm.java]: This method replaces the code above for resolving the code to an accessToken.

**Comment by samrocketman** [comment]
Please rebase this.  #61 has been merged.

**Comment by samrocketman** [comment]
@i386 I'm also having trouble with the `pom2.3` branch building with Java 1.8.  Do you mind taking a look at it as well?  The reason to migrate to pom2.3 is releasing via HTTPS.

**Comment by i386** [comment]
@samrocketman I've rebased and pushed - this PR is ready for review.

Strange that you are having issues on Java 8 as thats what I've been building it with locally. What error are you seeing?

**Comment by samrocketman** [comment]
I'll start reviewing this PR and merge if I think it's ready.

---
in reply to:
> Strange that you are having issues on Java 8 as thats what I've been building it with locally. What error are you seeing?

I'm using:
* Oracle java version "1.8.0_91".
* Apache Maven 3.3.9

On the `pom2.3` branch I get errors like:
```
testCanReadAndBuildOrgRepositoryICollaborateOn(org.jenkinsci.plugins.GithubRequireOrganizationMembershipACLTest)  Time elapsed: 0.003 sec  <<< ERROR!
java.lang.IllegalStateException: attempt to register a second Permission for hudson.model.Item.ViewStatus
        at hudson.security.PermissionGroup.add(PermissionGroup.java:72)
        at hudson.security.Permission.<init>(Permission.java:162)
        at hudson.security.Permission.<init>(Permission.java:168)
        at org.jenkinsci.plugins.GithubRequireOrganizationMembershipACLTest.<init>(GithubRequireOrganizationMembershipACLTest.java:95)
```
So I guess it's not specifically related to Java 8 in this case but due to upgrading the base version of Jenkins the plugin is built against.  I'm having trouble upgrading to the new Jenkins pom.xml format.  You can learn about what I tried in the commit message https://github.com/jenkinsci/github-oauth-plugin/commit/6a0aee1384f0d6ab64bb58e17ed4e486f2638d80.  There's also links I added as comments inline in the diff.

**Comment by samrocketman** [comment]
Tested OK.

**Pull request merged**